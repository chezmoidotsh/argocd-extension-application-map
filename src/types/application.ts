import { ArgoDestination, ArgoSource, ArgoSyncPolicy, Metadata } from './argocd';

/**
 * Type representing the possible application kinds
 */
export type ApplicationKind = 'Application' | 'ApplicationSet';

/**
 * Enum representing the possible health status values
 * @enum {string}
 */
export enum HealthStatus {
  /** Application is healthy and running as expected */
  Healthy = 'Healthy',
  /** Application is suspended */
  Suspended = 'Suspended',
  /** Application is in a degraded state */
  Degraded = 'Degraded',
  /** Application is currently progressing */
  Progressing = 'Progressing',
  /** Application is missing */
  Missing = 'Missing',
  /** Application health status is unknown */
  Unknown = 'Unknown',
}

/**
 * Converts a string to a HealthStatus enum value.
 * @param status - The string to convert.
 * @returns The HealthStatus enum value.
 */
export function StringToHealthStatus(str: string): HealthStatus {
  // Compatible with ES5/ES2015: use manual array
  const values = [
    HealthStatus.Healthy,
    HealthStatus.Suspended,
    HealthStatus.Degraded,
    HealthStatus.Progressing,
    HealthStatus.Missing,
    HealthStatus.Unknown,
  ];
  return values.indexOf(str as HealthStatus) !== -1 ? (str as HealthStatus) : HealthStatus.Unknown;
}

/**
 * Enum representing the possible sync status values
 * @enum {string}
 */
export enum SyncStatus {
  /** Application is in sync with the desired state */
  Synced = 'Synced',
  /** Application is out of sync with the desired state */
  OutOfSync = 'OutOfSync',
  /** Application sync status is unknown */
  Unknown = 'Unknown',
}

/**
 * Converts a string to a SyncStatus enum value.
 * @param status - The string to convert.
 * @returns The SyncStatus enum value.
 */
export function StringToSyncStatus(str: string): SyncStatus {
  const values = [SyncStatus.Synced, SyncStatus.OutOfSync, SyncStatus.Unknown];
  return values.indexOf(str as SyncStatus) !== -1 ? (str as SyncStatus) : SyncStatus.Unknown;
}

/**
 * Interface for application specification
 * @interface ApplicationSpec
 * @property {ArgoDestination} destination - The destination configuration for the application
 * @property {string} project - The ArgoCD project name
 * @property {ArgoSyncPolicy} syncPolicy - The synchronization policy configuration
 */
export interface ApplicationSpec {
  sources: ArgoSource[];
  destination: ArgoDestination;
  project: string;
  syncPolicy: ArgoSyncPolicy;
}

/**
 * Interface for application status
 * @interface ApplicationStatus
 * @property {HealthStatus} health - The current health status of the application
 * @property {SyncStatus} sync - The current sync status of the application
 */
export interface ApplicationStatus {
  health: HealthStatus;
  sync: SyncStatus;
}

/**
 * Interface representing a single ArgoCD application
 * @interface Application
 * @property {string} kind - The kind of the application, always "Application"
 * @property {Metadata} metadata - Application metadata including name, namespace, and labels
 * @property {ApplicationSpec} spec - Application specification including sources, destination, project, and sync policy
 * @property {ApplicationStatus} status - Current status including health and sync state
 */
export interface Application {
  kind: 'Application';
  metadata: Metadata;
  spec: ApplicationSpec;
  status: ApplicationStatus;
}

/**
 * Interface representing an ArgoCD ApplicationSet
 * @interface ApplicationSet
 * @property {string} kind - The kind of the application set, always "ApplicationSet"
 * @property {Metadata} metadata - Application set metadata including name, namespace, and labels
 */
export interface ApplicationSet {
  kind: 'ApplicationSet';
  metadata: Metadata;
}

/**
 * Interface representing an ArgoCD application or application set
 * @interface ApplicationUnion
 *
 * This is a discriminated union type with two variants:
 *
 * 1. Application:
 *    - Represents a single ArgoCD application
 *    - Contains a complete specification including sources, destination, project and sync policy
 *    - Used for managing individual application deployments
 *
 * 2. ApplicationSet:
 *    - Represents an ArgoCD ApplicationSet
 *    - Does not contain a spec as ApplicationSets are managed differently
 *    - Used for managing multiple applications through templates
 *
 * @property {ApplicationKind} kind - The kind of application (Application or ApplicationSet)
 * @property {Metadata} metadata - Application metadata including name, namespace, and labels
 * @property {ApplicationSpec} [spec] - Application specification (only present for Application kind)
 * @property {ApplicationStatus} [status] - Current status including health and sync state (only present for Application kind)
 */
export type ApplicationUnion = Application | ApplicationSet;
