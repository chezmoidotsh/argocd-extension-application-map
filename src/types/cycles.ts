/**
 * Types for cycle detection functionality in the application dependency graph
 */

/**
 * Information about a single edge in the dependency graph
 */
export interface EdgeInfo {
  /** Source node identifier */
  source: string;
  /** Target node identifier */
  target: string;
  /** Unique edge identifier (typically "source-target") */
  id: string;
}

/**
 * Information about a detected cycle in the dependency graph
 */
export interface CycleInfo {
  /** Unique identifier for the cycle (nodes joined by "->") */
  id: string;
  /** Array of node identifiers that form the cycle (ordered) */
  nodes: string[];
  /** Array of edges that form the cycle */
  edges: EdgeInfo[];
}

/**
 * Result of cycle detection algorithm
 */
export interface CycleDetectionResult {
  /** Whether any cycles were detected in the graph */
  hasCycle: boolean;
  /** Array of all detected cycles */
  cycles: CycleInfo[];
}
