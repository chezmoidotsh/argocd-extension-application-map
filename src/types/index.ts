export { ConnectionStatus } from './sse';
export { isApplication, isApplicationSet, HealthStatus, SyncStatus } from './application';
export { RankDirection } from './rankdirection';
export type { Application, ApplicationSet } from './application';
export type { RankDirectionType } from './rankdirection';
export type { SSEEvent, ConnectionStatusDetails } from './sse';
