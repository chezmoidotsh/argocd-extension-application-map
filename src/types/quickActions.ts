/**
 * Types for Quick Actions functionality
 */

/**
 * Enum representing the possible states of a quick action
 */
export enum QuickActionState {
  Idle = 'Idle',
  Loading = 'Loading',
  Success = 'Success',
  Error = 'Error',
}
