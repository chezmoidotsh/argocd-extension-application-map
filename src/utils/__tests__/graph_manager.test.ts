import { DirectedGraph } from 'graphology';

import { HealthStatus, SyncStatus } from '../../types/application';
import { ArgoApplication, ArgoApplicationSet } from '../../types/argocd';
import { updateGraph } from '../graph_manager';

// Helper to create a mock application payload
const createMockApp = (
  name: string,
  owner?: { kind: 'ApplicationSet'; name: string },
  resources?: any[]
): ArgoApplication => {
  const app: ArgoApplication = {
    kind: 'Application',
    metadata: {
      name,
      namespace: 'argocd',
      ...(owner
        ? {
            ownerReferences: [
              {
                apiVersion: 'argoproj.io/v1alpha1',
                kind: owner.kind,
                name: owner.name,
              },
            ],
          }
        : {}),
    },
    status: {
      health: { status: HealthStatus.Healthy },
      sync: { status: SyncStatus.Synced },
      resources: resources || [],
    },
  };
  return app;
};

// --- TESTS ---
describe('updateGraph', () => {
  /**
   * Diagram:
   *   ┌────────────┐
   *   │ appA       │
   *   └────────────┘
   * (Single node added)
   */
  it('adds a simple application node', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const appA = createMockApp('appA');
    updateGraph(graph, 'ADDED', appA);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Diagram:
   *   ┌────────────┐     ┌──────────────┐
   *   │ appset1    │ ◁── │ appB         │
   *   └────────────┘     └──────────────┘
   * (appB owned by appset1)
   */
  it('adds a node with ownerReferences', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const app = createMockApp('appB', { kind: 'ApplicationSet', name: 'appset1' });
    updateGraph(graph, 'ADDED', app);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Diagram:
   *   ┌────────────┐
   *   │ appA       │
   *   └───┬────────┘
   *       │
   *   ┌───▽────────┐
   *   │ appB       │
   *   └────────────┘
   * (appA references appB in status.resources)
   */
  it('adds resources from status.resources as stub nodes', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const app = createMockApp('appA', undefined, [{ kind: 'Application', name: 'appB', namespace: 'argocd' }]);
    updateGraph(graph, 'ADDED', app);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Initial diagram:
   *   ┌────────────┐
   *   │ appA       │
   *   └────────────┘
   * After MODIFIED:
   *   ┌────────────┐
   *   │ appA       │ (status updated)
   *   └────────────┘
   */
  it('updates the attributes of an existing node', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const app = createMockApp('appA');
    updateGraph(graph, 'ADDED', app);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Initial diagram:
   *   ┌────────────┐
   *   │ appA       │
   *   └────────────┘
   * After DELETED:
   *   (no more appA node)
   */
  it('removes a node on DELETED', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const app = createMockApp('appA');
    updateGraph(graph, 'ADDED', app);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Diagram:
   *   ┌────────────┐     ┌──────────────┐
   *   │ appA       │ ──▶ │ appset1      │
   *   └────────────┘     └──────────────┘
   * (appA references appset1 in status.resources)
   */
  it('adds an ApplicationSet referenced in resources', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const app = createMockApp('appA', undefined, [{ kind: 'ApplicationSet', name: 'appset1', namespace: 'argocd' }]);
    updateGraph(graph, 'ADDED', app);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Diagram:
   *   ┌────────────┐     ┌──────────────┐     ┌──────────────┐
   *   │ app-of-apps│ ──▶ │ appA         │ ──▶ │ appB         │
   *   └────────────┘     └──────────────┘     └──────────────┘
   * (appA references appB in status.resources)
   */
  it('should handle app-of-apps without ownerReferences', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const appA = createMockApp('appA', { kind: 'ApplicationSet', name: 'app-of-apps' }, [
      { kind: 'Application', name: 'appB', namespace: 'argocd' },
    ]);
    const appB = createMockApp('appB');

    updateGraph(graph, 'ADDED', appA);
    updateGraph(graph, 'ADDED', appB);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Initial diagram:
   *   ┌────────────┐     ┌──────────────┐
   *   │ appset1    │ ◁── │ appA         │
   *   └────────────┘     └──────────────┘
   *   ┌────────────┐     ┌──────────────┐
   *   │ appset1    │ ◁── │ appB         │
   *   └────────────┘     └──────────────┘
   * After DELETED appA:
   *   ┌────────────┐     ┌──────────────┐
   *   │ appset1    │ ◁── │ appB         │
   *   └────────────┘     └──────────────┘
   * After DELETED appB:
   *   (appset1 is removed as it has no children)
   */
  it('removes an ApplicationSet parent when it becomes orphaned', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const appA = createMockApp('appA', { kind: 'ApplicationSet', name: 'appset1' });
    const appB = createMockApp('appB', { kind: 'ApplicationSet', name: 'appset1' });

    updateGraph(graph, 'ADDED', appA);
    updateGraph(graph, 'ADDED', appB);

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Initial diagram:
   *   ┌────────────┐     ┌────────────┐
   *   │ appA       │ ──▶ │ appB       │
   *   └────────────┘     └────────────┘
   * After MODIFIED appA:
   *   ┌────────────┐     ┌────────────┐
   *   │ appA       │ ──▶ │ appC       │
   *   └────────────┘     └────────────┘
   *   ┌────────────┐
   *   │ appB       │
   *   └────────────┘
   */
  it('updates resources for a node on MODIFIED event', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const appA_v1 = createMockApp('appA', undefined, [{ kind: 'Application', name: 'appB', namespace: 'argocd' }]);
    updateGraph(graph, 'ADDED', appA_v1);
    expect(graph.export()).toMatchSnapshot();

    const appA_v2 = createMockApp('appA', undefined, [{ kind: 'Application', name: 'appC', namespace: 'argocd' }]);
    updateGraph(graph, 'MODIFIED', appA_v2);

    expect(graph.hasEdge('Application/argocd/appA', 'Application/argocd/appB')).toBe(false);
    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Diagram:
   *   ┌────────────┐     ┌────────────┐
   *   │ appA       │ ──▶ │ appB       │
   *   └────────────┘     └────────────┘
   *                      (status: Unknown/Unknown)
   *
   * After appA is processed, which lists appB as Healthy/Synced in its resources,
   * appB's status in the graph should be updated.
   *
   *   ┌────────────┐     ┌────────────┐
   *   │ appA       │ ──▶ │ appB       │
   *   └────────────┘     └────────────┘
   *                      (status: Healthy/Synced)
   */
  it("updates a child's unknown status from a parent's resource definition", () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();

    const appA = createMockApp('appA', undefined, [
      {
        kind: 'Application',
        name: 'appB',
        namespace: 'argocd',
        health: { status: HealthStatus.Unknown },
        status: SyncStatus.Unknown,
      },
    ]);
    updateGraph(graph, 'ADDED', appA);

    expect(graph.export()).toMatchSnapshot();

    // 2. Create appA, which lists appB as a resource with a known status
    updateGraph(graph, 'MODIFIED', {
      ...appA,
      status: {
        ...appA.status,
        resources: appA.status.resources.map((resource) => ({
          ...resource,
          health: { status: HealthStatus.Healthy },
          status: SyncStatus.Synced,
        })),
      },
    });

    expect(graph.export()).toMatchSnapshot();
  });

  /**
   * Initial diagram:
   *   ┌────────────┐     ┌──────────────┐     ┌────────────┐
   *   │ appset1    │ ◁── │ appA         │ ──▶ │ appB       │
   *   └────────────┘     └──────────────┘     └────────────┘
   * After DELETED appA:
   *                                              ┌────────────┐
   *                                              │ appB       │
   *                                              └────────────┘
   *   (appset1 is removed as it's orphaned, appB remains)
   */
  it('correctly handles deletion of a node that is both a parent and a child', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const appA = createMockApp('appA', { kind: 'ApplicationSet', name: 'appset1' }, [
      { kind: 'Application', name: 'appB', namespace: 'argocd' },
    ]);
    const appB = createMockApp('appB');

    updateGraph(graph, 'ADDED', appB);
    updateGraph(graph, 'ADDED', appA);

    expect(graph.export()).toMatchSnapshot();

    updateGraph(graph, 'DELETED', appA);

    expect(graph.export()).toMatchSnapshot();
  });

  it('handles incomplete or malformed SSE payloads gracefully', () => {
    const graph = new DirectedGraph<ArgoApplication | ArgoApplicationSet>();
    const malformedApp: ArgoApplication = {
      kind: 'Application',
      metadata: {
        name: 'malformedApp',
        namespace: 'argocd',
      },
      status: {
        health: { status: undefined },
        sync: { status: undefined },
        resources: [],
      },
    };

    updateGraph(graph, 'ADDED', malformedApp);
    expect(graph.export()).toMatchSnapshot();

    const appWithBadResource = createMockApp('appWithBadResource', undefined, [
      { kind: 'Application', name: null, namespace: 'argocd' }, // name is null
    ]);
    updateGraph(graph, 'ADDED', appWithBadResource);
    expect(graph.export()).toMatchSnapshot();
  });
});
