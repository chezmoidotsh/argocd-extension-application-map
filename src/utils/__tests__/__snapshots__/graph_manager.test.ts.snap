// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`updateGraph adds a node with ownerReferences 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "ApplicationSet/argocd/appset1->Application/argocd/appB",
      "source": "ApplicationSet/argocd/appset1",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
          "ownerReferences": [
            {
              "apiVersion": "argoproj.io/v1alpha1",
              "kind": "ApplicationSet",
              "name": "appset1",
            },
          ],
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
    {
      "attributes": {
        "kind": "ApplicationSet",
        "metadata": {
          "name": "appset1",
          "namespace": "argocd",
        },
      },
      "key": "ApplicationSet/argocd/appset1",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph adds a simple application node 1`] = `
{
  "attributes": {},
  "edges": [],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph adds an ApplicationSet referenced in resources 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "Application/argocd/appA->ApplicationSet/argocd/appset1",
      "source": "Application/argocd/appA",
      "target": "ApplicationSet/argocd/appset1",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "ApplicationSet",
              "name": "appset1",
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "ApplicationSet",
        "metadata": {
          "name": "appset1",
          "namespace": "argocd",
        },
        "status": {
          "health": undefined,
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "ApplicationSet/argocd/appset1",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph adds resources from status.resources as stub nodes 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "Application/argocd/appA->Application/argocd/appB",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "Application",
              "name": "appB",
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": undefined,
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph correctly handles deletion of a node that is both a parent and a child 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "ApplicationSet/argocd/appset1->Application/argocd/appA",
      "source": "ApplicationSet/argocd/appset1",
      "target": "Application/argocd/appA",
    },
    {
      "key": "Application/argocd/appA->Application/argocd/appB",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
          "ownerReferences": [
            {
              "apiVersion": "argoproj.io/v1alpha1",
              "kind": "ApplicationSet",
              "name": "appset1",
            },
          ],
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "Application",
              "name": "appB",
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "ApplicationSet",
        "metadata": {
          "name": "appset1",
          "namespace": "argocd",
        },
      },
      "key": "ApplicationSet/argocd/appset1",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph correctly handles deletion of a node that is both a parent and a child 2`] = `
{
  "attributes": {},
  "edges": [],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph handles incomplete or malformed SSE payloads gracefully 1`] = `
{
  "attributes": {},
  "edges": [],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "malformedApp",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": undefined,
          },
          "resources": [],
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "Application/argocd/malformedApp",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph handles incomplete or malformed SSE payloads gracefully 2`] = `
{
  "attributes": {},
  "edges": [],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "malformedApp",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": undefined,
          },
          "resources": [],
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "Application/argocd/malformedApp",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appWithBadResource",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "Application",
              "name": null,
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appWithBadResource",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph removes a node on DELETED 1`] = `
{
  "attributes": {},
  "edges": [],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph removes an ApplicationSet parent when it becomes orphaned 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "ApplicationSet/argocd/appset1->Application/argocd/appA",
      "source": "ApplicationSet/argocd/appset1",
      "target": "Application/argocd/appA",
    },
    {
      "key": "ApplicationSet/argocd/appset1->Application/argocd/appB",
      "source": "ApplicationSet/argocd/appset1",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
          "ownerReferences": [
            {
              "apiVersion": "argoproj.io/v1alpha1",
              "kind": "ApplicationSet",
              "name": "appset1",
            },
          ],
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "ApplicationSet",
        "metadata": {
          "name": "appset1",
          "namespace": "argocd",
        },
      },
      "key": "ApplicationSet/argocd/appset1",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
          "ownerReferences": [
            {
              "apiVersion": "argoproj.io/v1alpha1",
              "kind": "ApplicationSet",
              "name": "appset1",
            },
          ],
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph should handle app-of-apps without ownerReferences 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "ApplicationSet/argocd/app-of-apps->Application/argocd/appA",
      "source": "ApplicationSet/argocd/app-of-apps",
      "target": "Application/argocd/appA",
    },
    {
      "key": "Application/argocd/appA->Application/argocd/appB",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
          "ownerReferences": [
            {
              "apiVersion": "argoproj.io/v1alpha1",
              "kind": "ApplicationSet",
              "name": "app-of-apps",
            },
          ],
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "Application",
              "name": "appB",
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "ApplicationSet",
        "metadata": {
          "name": "app-of-apps",
          "namespace": "argocd",
        },
      },
      "key": "ApplicationSet/argocd/app-of-apps",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph updates a child's unknown status from a parent's resource definition 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "Application/argocd/appA->Application/argocd/appB",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "health": {
                "status": "Unknown",
              },
              "kind": "Application",
              "name": "appB",
              "namespace": "argocd",
              "status": "Unknown",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Unknown",
          },
          "sync": {
            "status": "Unknown",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph updates a child's unknown status from a parent's resource definition 2`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "Application/argocd/appA->Application/argocd/appB",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "health": {
                "status": "Healthy",
              },
              "kind": "Application",
              "name": "appB",
              "namespace": "argocd",
              "status": "Synced",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph updates resources for a node on MODIFIED event 1`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "Application/argocd/appA->Application/argocd/appB",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appB",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "Application",
              "name": "appB",
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": undefined,
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "Application/argocd/appB",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph updates resources for a node on MODIFIED event 2`] = `
{
  "attributes": {},
  "edges": [
    {
      "key": "Application/argocd/appA->Application/argocd/appC",
      "source": "Application/argocd/appA",
      "target": "Application/argocd/appC",
    },
  ],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [
            {
              "kind": "Application",
              "name": "appC",
              "namespace": "argocd",
            },
          ],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appB",
          "namespace": "argocd",
        },
        "status": {
          "health": undefined,
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "Application/argocd/appB",
    },
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appC",
          "namespace": "argocd",
        },
        "status": {
          "health": undefined,
          "sync": {
            "status": undefined,
          },
        },
      },
      "key": "Application/argocd/appC",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;

exports[`updateGraph updates the attributes of an existing node 1`] = `
{
  "attributes": {},
  "edges": [],
  "nodes": [
    {
      "attributes": {
        "kind": "Application",
        "metadata": {
          "name": "appA",
          "namespace": "argocd",
        },
        "status": {
          "health": {
            "status": "Healthy",
          },
          "resources": [],
          "sync": {
            "status": "Synced",
          },
        },
      },
      "key": "Application/argocd/appA",
    },
  ],
  "options": {
    "allowSelfLoops": true,
    "multi": false,
    "type": "directed",
  },
}
`;
